plugins/plugin-machine/che-plugin-machine-ext-server/src/main/java/org/eclipse/che/ide/ext/machine/server/ssh/KeysInjector.java

```diff
@@ -66,15 +67,41 @@ public void start() {
             @Override
             public void onEvent(MachineStatusEvent event) {
                 if (event.getEventType() == MachineStatusEvent.EventType.RUNNING) {
+                    final Instance machine;
                     try {
-                        final Instance machine = environmentEngine.getMachine(event.getWorkspaceId(),
-                                                                              event.getMachineId());
+                        machine = environmentEngine.getMachine(event.getWorkspaceId(),
+                                                               event.getMachineId());
+                    } catch (NotFoundException e) {
+                        LOG.error("Unable to find machine: " + e.getLocalizedMessage(), e);
+                        return;
+                    }
+
+                    try {
+                        // get machine keypairs
                         List<SshPairImpl> sshPairs = sshManager.getPairs(machine.getOwner(), "machine");
-                        final List<String> publicKeys = sshPairs.stream()
+                        final List<String> publicMachineKeys = sshPairs.stream()
                                                              .filter(sshPair -> sshPair.getPublicKey() != null)
                                                              .map(SshPairImpl::getPublicKey)
                                                              .collect(Collectors.toList());
 
+                        // get workspace keypair (if any)
+                        SshPairImpl sshWorkspacePair = null;
+                        try {
+                            sshWorkspacePair = sshManager.getPair(machine.getOwner(), "workspace", event.getWorkspaceId());
+                        } catch (NotFoundException e) {
+                            LOG.debug("No ssh key associated to the workspace", e);
+                        }
+
+                        // build list of all pairs.
+                        final List<String> publicKeys;
+                        if (sshWorkspacePair != null && sshWorkspacePair.getPublicKey() != null) {
+                            publicKeys = new ArrayList<>(publicMachineKeys.size() + 1);
```

Comment:
Why should we create new list? Why we can't just add sshWorkspacePair to publicMachineKeys list?